rust/connlib/clients/shared/src/eventloop.rs

```diff
@@ -64,6 +65,12 @@ where
         loop {
             match self.rx.poll_recv(cx) {
                 Poll::Ready(Some(Command::Stop)) | Poll::Ready(None) => return Poll::Ready(Ok(())),
+                Poll::Ready(Some(Command::Reconnect)) => {
```

Comment:
I wonder if using a channel for this is the most convenient way to go about this.

We might want to use a different mechanism so that multiple `reconnects` aren't queued up I was thinking we can use a `Notify`, that way we don't need to worry about the bounded channel and there's no point on doing multiple reconnects in a row we want to just listen to the latest.

Reply:
Yeah if they're guaranteed idempotent that would keep the channel from filling up. We did a similar thing for `on_update_resources` in the Tauri Client: https://github.com/firezone/firezone/blob/ffc034d5c4d0224248346c72f5272637403cc100/rust/gui-client/src-tauri/src/client/gui.rs#L484

It was possible if I only allowed 5 items in the channel, and connlib rapidly sent `on_update_resources` events, that the channel might fill up and error (since it's not allowed to block the callbacks) before the GUI got around to dealing with them.

So the channel was replaced with a `Notify` and something that the reader can poll when it's notified, same as if it was a channel that dropped all but the most recent event.

I think we also considered tokio's `watch` and it wasn't a perfect fit. `Notify` has worked well.

Reply:
Unfortunately, `Notify` doesn't have a `poll` API so it would be a bit clunky to use. If debouncing is what we want, then I can add a small delay to the sending of the command through the channel and cancel the current send if we get another one.

Reply:
Huh, yeah it doesn't. And it can't be replicated with `AtomicWaker`?

Reply:
hm, a size 1 channel would also achieve the same effect as `Notify`

Reply:
> hm, a size 1 channel would also achieve the same effect as `Notify`

We already only have a size-1 channel so we drop repeated `Reconnect`s if they eventloop hasn't processed it yet. I'd be surprised though if the app can actually call that quicker than our `Eventloop` can process it.

If we want debouncing, I'd say we probably want it on a scale of 500ms or something like that.

Reply:
Shall we ship this for now and if we notice that there are too many subsequent reconnect calls, we add debouncing?

Reply:
Hm so it's a 1-sized channel that uses `try_send` for both commands, so if I did `reconnect` and `stop` in the same tick somehow, the `stop` will be silently ignored?

Reply:
Yes. The `Stop` isn't super critical though. If you drop `Session`, the `Runtime` gets dropped and with it, all tasks should be stopped.

Reply:
Even the `Stop` does not need a command, as much as a "I want you to be running / not be running" flag and a way to notify when it's changed.

With channels we have to trade off between 3 problems: Sender may block, sends may fail silently, or sends may panic.

Reply:
Are you suggesting using shared memory instead of channels instead and just notify when to re-read the shared memory?

Reply:
Kinda. `AtomicBool`, so not like it has to lock.
But I wrote my last comment at the same time as you wrote "The Stop isn't super critical", so it might just be something to put as an issue and merge this PR anyway